src/_purryOrderRules.ts

```diff
@@ -0,0 +1,147 @@
+import type { NonEmptyArray } from './_types';
+
+export type OrderRule<T> = OrderProjection<T> | OrderPair<T>;
+export type CompareFunction<T> = (a: T, b: T) => number;
+
+type OrderPair<T> = readonly [
+  projector: OrderProjection<T>,
+  direction: Direction,
+];
+
+type OrderProjection<T> = (x: T) => Comparable;
+
+type Comparable = ComparablePrimitive | { valueOf(): ComparablePrimitive };
+type ComparablePrimitive = number | string | boolean;
+
+const ALL_DIRECTIONS = ['asc', 'desc'] as const;
+type Direction = (typeof ALL_DIRECTIONS)[number];
```

Comment:
**question:** is there a reason it's in this order?

i found it a bit confusing to read for the first time, maybe reordering would help (similar to the order in old `sortBy`)

(comments might be overkill, but whatever)

```ts
type ComparablePrimitive = number | string | boolean;

/** A type that we can compare with < or >. */
type Comparable = ComparablePrimitive | { valueOf(): ComparablePrimitive };

type OrderProjection<T> = (x: T) => Comparable;

const ALL_DIRECTIONS = ['asc', 'desc'] as const;
type Direction = (typeof ALL_DIRECTIONS)[number];

type OrderPair<T> = readonly [
  projector: OrderProjection<T>,
  direction: Direction,
];

/** A function, or a function and a direction, used to give a key to order by. */
export type OrderRule<T> = OrderProjection<T> | OrderPair<T>;

export type CompareFunction<T> = (a: T, b: T) => number;
```

Reply:
I usually put exported stuff at the top and implementation details below (I also do this with functions). That way, people have to scroll down less to see the primary focus of a file. Anything that isn't exported is implementation details that shouldn't matter to anyone except maintainers.